From 7f7ecb67bc0e07acb644081e5834f34f5fa3f4b0 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 29 Sep 2021 08:21:20 +0300
Subject: [PATCH] [SW-57369] gaudi2: arc: add sram range check

This patch adds platform specific SRAM range check function
as the SRAM address received needs to be converted to ARC viewpoint
and then checked for valid range.

Change-Id: I91032bf5e6398d26a4ea19f0b41d57638a7b5667
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 9623409..971e026 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -392,6 +392,21 @@
 
 	preboot_puts_hex64("ERROR: invalid address range addr = ",
 			   addr, true);
+
+	return false;
+}
+
+bool preboot_sram_is_valid_address_range(u64 addr, int len)
+{
+	/* Prepare the ARC SRAM address */
+	u32 arc_addr = ((u32)addr & ARM_ADDR_MASK) | BIT(28);
+
+	if (arc_addr >= SDRAM_BASE && arc_addr + len <= SDRAM_BASE + SRAM_SIZE)
+		return !preboot_is_overlap_address_range(arc_addr, len);
+
+	if (addr >= FIT_IMG_SPSRAM_BASE && addr + len <= FIT_IMG_SPSRAM_END)
+		return true;
+
 	return false;
 }
 
